### PR TITLE
Add example for co-existence of klog v1 and v2

### PR DESCRIPTION
**Pull request by dims**
Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What this PR does / why we need it**:
Now that we have v1 and v2, we need to show folks how they can deal with a codebase where both are present until dependencies can move to klog v2.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Example to show how v1 and v2 can co-exist.
```

### CODE REVIEW

**Comment by dims** [comment]
Sample output:
```
Log file created at: 2020/05/14 16:56:53
Running on machine: dims-a02
Binary: Built with gc go1.13.10 for darwin/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0514 16:56:53.269981   34750 coexist_klog_v1_and_v2.go:60] hello from klog v1!
I0514 16:56:53.270250   34750 coexist_klog_v1_and_v2.go:61] go:61] hello from klog v2!
W0514 16:56:53.270551   34750 coexist_klog_v1_and_v2.go:62] go:62] beware from klog v2!
E0514 16:56:53.270829   34750 coexist_klog_v1_and_v2.go:63] go:63] error from klog v2!
I0514 16:56:53.270832   34750 coexist_klog_v1_and_v2.go:64] nice to meet you
```

**Comment by dims** [comment]
Normally we won't touch a branch like this ... this PR is *just* an example. We need a official place for this example as we need to throw this into release notes for 1.19 and possibly a blog post etc.

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	} else if p[0] == 'W' {
+		klogv2.WarningDepth(OutputCallDepth, string(p[DefaultPrefixLength:]))
+	} else if p[0] == 'E' {
+		klogv2.ErrorDepth(OutputCallDepth, string(p[DefaultPrefixLength:]))
+	} else if p[0] == 'F' {
+		klogv2.FatalDepth(OutputCallDepth, string(p[DefaultPrefixLength:]))
+	} else {
+		klogv2.InfoDepth(OutputCallDepth, string(p[DefaultPrefixLength:]))
+	}
+	return len(p) - DefaultPrefixLength, nil
```
Comment (liggitt) [inline examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:35]: we handled the full length, so return len(p)

From the writer interface:

> Write must return a non-nil error if it returns n < len(p).

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	// These following are just things you do to ensure that you stream
+	// all the things to the log file instead of stdout/stderr
+	flag.Set("logtostderr", "false")     // By default klog logs to stderr, switch that off
+	flag.Set("alsologtostderr", "false") // false is default, but this is informative
+	flag.Set("stderrthreshold", "FATAL") // stderrthreshold defaults to ERROR, we don't want anything in stderr
+	flag.Set("log_file", "myfile.log")   // log to a file
+	flag.Parse()
+
+	// BEGIN : hack to redirect klogv1 calls to klog v2
+	// Tell klog to log into STDERR. Otherwise, we risk
```
Comment (liggitt) [inline examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:51]: *not* to log into stderr?

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	// certain kinds of API errors getting logged into a directory not
+	// available in a `FROM scratch` Docker container, causing us to abort
+	var flags flag.FlagSet
```
Comment (liggitt) [inline examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:54]: clarify these are the flags bound to klog v1?
```suggestion
	var klog1Flags flag.FlagSet
```

**Comment by neolit123** [comment]
from my experience, co-existence examples and docs for v1+v2 of a project reside in the v2 version.
maybe that's not possible here / shrug.

**Comment by dims** [comment]
> from my experience, co-existence examples and docs for v1+v2 of a project reside in the v2 version.
> maybe that's not possible here / shrug.

cc @liggitt ^^^

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	// Assume that we have to switch klogv2 on from now on everywhere.
+	klogv2.InitFlags(nil)
+
+	// These following are just things you do to ensure that you stream
+	// all the things to the log file instead of stdout/stderr
+	flag.Set("logtostderr", "false")     // By default klog logs to stderr, switch that off
+	flag.Set("alsologtostderr", "false") // false is default, but this is informative
+	flag.Set("stderrthreshold", "FATAL") // stderrthreshold defaults to ERROR, we don't want anything in stderr
+	flag.Set("log_file", "myfile.log")   // log to a file
+	flag.Parse()
```
Comment (liggitt) [inline examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:48]: are all of these options required for this example? I thought we were just 

can we simplify the klog/v2 portion to:

```go
// initialize klog/v2, can also bind to a local flagset if desired
klogv2.InitFlags(nil)
// ... add any other desired klog/v2 setup here
// parse klog/v2 flags
flag.Parse()
// make sure we flush before exiting
defer klogv2.Flush()
```

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	klogv1.SetOutputBySeverity("INFO", klogWriter{})
+	// END : hack to redirect klogv1 calls to klog v2
+
+	// Now you can mix klogv1 and v2 in the same code base
+	klogv2.Info("hello from klog (v2)!")
+	klogv1.Info("hello from klog (v1)!")
+	klogv1.Warning("beware from klog (v1)!")
+	klogv1.Error("error from klog (v1)!")
+	klogv2.Info("nice to meet you (v2)")
+	klogv2.Flush()
```
Comment (liggitt) [inline examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:66]: drop the flush if we deferred it above?

**Comment by liggitt** [comment]
> > from my experience, co-existence examples and docs for v1+v2 of a project reside in the v2 version.
> > maybe that's not possible here / shrug.
> 
> cc @liggitt ^^^

I guess it doesn't particularly matter which branch the example lives in... thing I care most about is demonstrating v1 -> v2 routing, not the reverse

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go (hunk 1 above)
Comment (dims) [reply examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:35]: Fixed.

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go (hunk 2 above)
Comment (dims) [reply examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:51]: Done

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go (hunk 3 above)
Comment (dims) [reply examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:54]: Done

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go (hunk 4 above)
Comment (dims) [reply examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:48]: I'd like to keep them for illustration purposes when folks try out the example, they should see nothing in stderr/out and all the log lines should be in the myfile.log, i'll try to word the comments better.

File: examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go (hunk 5 above)
Comment (dims) [reply examples/coexist_klog_v1_and_v2/coexist_klog_v1_and_v2.go:66]: done

**Comment by dims** [comment]
I've incorporated the suggestions, will open another PR against master

**Pull request closed without merging**